sale_coupon_limit/models/sale_coupon.py

```diff
@@ -0,0 +1,66 @@
+# Copyright 2021 Tecnativa - David Vidal
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from odoo import _, models
+
+
+class SaleCoupon(models.Model):
+    _inherit = "sale.coupon"
+
+    def _check_coupon_code(self, order):
+        """Add customer and salesmen limit to program coupons. Check the error strings
+        for a detailed case detail."""
+        message = super()._check_coupon_code(order)
+        if message:
+            return message
+        salesman_rule = (
+            order.user_id
+            and self.program_id
+            and self.program_id.rule_salesmen_limit_ids.filtered(
+                lambda x: order.user_id in x.rule_user_ids
+            )[:1]
+        )
+        # Customer limit rules
+        if self.program_id and self.program_id.rule_max_customer_application:
+            coupons_count = self.search_count(
+                [
+                    ("order_id", "!=", order.id),
+                    (
+                        "sales_order_id.partner_id",
+                        "child_of",
+                        order.partner_id.commercial_partner_id.id,
```

Comment:
This search is not correct if you have a contact in the order. It should be:
```suggestion
                        "sales_order_id.partner_id.commercial_partner_id",
                        "=",
                        order.partner_id.commercial_partner_id.id,
```
Maybe it's interesting to depend on `sale_commercial_partner` for having this field at order level if we want good performance.